plugin/metrics/metrics.go

```diff
@@ -95,35 +97,46 @@ func (m *Metrics) OnStartup() error {
 	m.mux = http.NewServeMux()
 	m.mux.Handle("/metrics", promhttp.HandlerFor(m.Reg, promhttp.HandlerOpts{}))
 
+	m.srv = &http.Server{Handler: m.mux}
 	go func() {
-		http.Serve(m.ln, m.mux)
+		m.srv.Serve(m.ln)
 	}()
+
 	return nil
 }
 
 // OnRestart stops the listener on reload.
-func (m *Metrics) OnRestart() error {
+func (m *Metrics) stopService() {
+	// We allow prometheus statements in multiple Server Blocks, but only the first
+	// will open the listener, for the rest they are all nil; guard against that.
+
 	if !m.lnSetup {
-		return nil
+		return
 	}
 
-	uniqAddr.SetTodo(m.Addr)
+	// Shutdown will ensure that the service is completely stopped - needed for some tests.
+	err := m.srv.Shutdown(context.TODO())
+	if err != nil {
+		log.Infof("Error when closing prometheus http server: %s", err)
+	}
 
-	m.ln.Close()
 	m.lnSetup = false
+	log.Infof("listener stopped on %s", m.Addr)
+}
+
+func (m *Metrics) OnRestart() error {
+	m.stopService()
+
+	// un-register so a new metrics from the new config can register and be launched
+	uniqAddr.Unset(m.Addr)
+
```

Comment:
Here is the change to fix the initial issue : the current plugin is ended. So let's just unset from uniqAddr and let any new plugin setup to register for the next launching.
(and if there is no "prometheus" in the new corefile, then no listener will be launched)